File: steps/InnerHarness.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 25
//DEPS com.fasterxml.jackson.core:jackson-databind:2.17.2

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.ConnectException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;

// InnerHarness.java — shared library for Ollama-based local coding agents.
// Used via //SOURCES InnerHarness.java.

class InnerHarness {

    static final int MAX_OUTPUT = 10_000;
    static final ObjectMapper JSON = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
            .setSerializationInclusion(JsonInclude.Include.NON_NULL);
    static final String DEFAULT_MODEL = "qwen3:8b";
    static final String DEFAULT_BASE_URL = "http://localhost:11434";

    static String modelName() {
        return System.getenv().getOrDefault("OLLAMA_MODEL", DEFAULT_MODEL);
    }

    static String baseUrl() {
        var value = System.getenv().getOrDefault("OLLAMA_BASE_URL", DEFAULT_BASE_URL);
        return value.endsWith("/") ? value.substring(0, value.length() - 1) : value;
    }

    static String setupHint() {
        var model = modelName();
        return "Start Ollama and pull the default model:\n"
                + "  ollama serve\n"
                + "  ollama pull " + model;
    }

    // Self-test when run directly
    void main(String[] args) {
        var agent = new Agent(25);
        var task = args.length > 0 ? String.join(" ", args)
                : "Read bookshelf/pom.xml and tell me what project this is.";
        agent.run(task);
    }

    static class ChatMessage {
        public String role;
        public String content;
        public String tool_name;
        public List<OllamaToolCall> tool_calls;

        static ChatMessage system(String content) { return of("system", content); }
        static ChatMessage user(String content) { return of("user", content); }
        static ChatMessage assistant(String content, List<OllamaToolCall> toolCalls) {
            var message = of("assistant", content == null ? "" : content);
            message.tool_calls = toolCalls == null || toolCalls.isEmpty() ? null : toolCalls;
            return message;
        }
        static ChatMessage tool(String toolName, String content) {
            var message = new ChatMessage();
            message.role = "tool";
            message.content = content;
            message.tool_name = toolName;
            return message;
        }
        static ChatMessage of(String role, String content) {
            var message = new ChatMessage();
            message.role = role;
            message.content = content;
            return message;
        }
    }

    static class OllamaFunctionCall {
        public String name;
        public JsonNode arguments;
    }

    static class OllamaToolCall {
        public OllamaFunctionCall function;
        String name() { return function == null ? "" : function.name; }
        JsonNode arguments() {
            return function == null || function.arguments == null ? JSON.createObjectNode() : function.arguments;
        }
        <T> T input(Class<T> type) {
            try { return JSON.treeToValue(arguments(), type); }
            catch (JsonProcessingException e) {
                throw new IllegalArgumentException("Invalid tool arguments for " + name() + ": " + e.getMessage(), e);
            }
        }
    }

    static class ChatResponse {
        public ChatMessage message;
        public String error;
        public Long total_duration;
        public Long load_duration;
        public Long prompt_eval_count;
        public Long prompt_eval_duration;
        public Long eval_count;
        public Long eval_duration;
    }

    record ToolSpec(String name, String description, Map<String, Object> parameters) {
        Map<String, Object> asOllamaTool() {
            return Map.of("type", "function", "function", Map.of(
                    "name", name,
                    "description", description,
                    "parameters", parameters));
        }
    }

    static Map<String, Object> objectSchema(Map<String, Object> properties, List<String> required) {
        var schema = new LinkedHashMap<String, Object>();
        schema.put("type", "object");
        schema.put("properties", properties);
        schema.put("required", required);
        return schema;
    }

    static Map<String, Object> stringProperty(String description) {
        return Map.of("type", "string", "description", description);
    }

    static class OllamaClient {
        final HttpClient httpClient;

        OllamaClient() {
            this.httpClient = HttpClient.newBuilder()
                    .connectTimeout(Duration.ofSeconds(10))
                    .build();
        }

        ChatResponse chat(List<ChatMessage> messages, List<ToolSpec> tools) {
            var body = new LinkedHashMap<String, Object>();
            body.put("model", modelName());
            body.put("stream", false);
            body.put("messages", messages);
            if (tools != null && !tools.isEmpty()) {
                body.put("tools", tools.stream().map(ToolSpec::asOllamaTool).toList());
            }

            try {
                var request = HttpRequest.newBuilder()
                        .uri(URI.create(baseUrl() + "/api/chat"))
                        .timeout(Duration.ofMinutes(5))
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(JSON.writeValueAsString(body)))
                        .build();
                var response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
                if (response.statusCode() < 200 || response.statusCode() >= 300) {
                    throw new IllegalStateException("Ollama chat request failed with HTTP "
                            + response.statusCode() + ": " + errorText(response.body()) + "\n\n" + setupHint());
                }

                var chatResponse = JSON.readValue(response.body(), ChatResponse.class);
                if (chatResponse.error != null && !chatResponse.error.isBlank()) {
                    throw new IllegalStateException("Ollama returned an error: " + chatResponse.error
                            + "\n\n" + setupHint());
                }
                if (chatResponse.message == null) {
                    throw new IllegalStateException("Ollama response did not include a message.\n\n" + setupHint());
                }
                return chatResponse;
            } catch (ConnectException e) {
                throw new IllegalStateException("Could not connect to Ollama at " + baseUrl() + ".\n\n" + setupHint(), e);
            } catch (IOException e) {
                if (isConnectException(e)) {
                    throw new IllegalStateException("Could not connect to Ollama at " + baseUrl() + ".\n\n" + setupHint(), e);
                }
                throw new IllegalStateException("Ollama chat request failed: " + e.getMessage()
                        + "\n\n" + setupHint(), e);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException("Interrupted while waiting for Ollama.\n\n" + setupHint(), e);
            }
        }

        static boolean isConnectException(Throwable throwable) {
            while (throwable != null) {
                if (throwable instanceof ConnectException) return true;
                throwable = throwable.getCause();
            }
            return false;
        }

        static String errorText(String body) {
            if (body == null || body.isBlank()) return "<empty response body>";
            try {
                var node = JSON.readTree(body);
                if (node.hasNonNull("error")) return node.get("error").asText();
            } catch (JsonProcessingException ignored) {
                // Return raw body below.
            }
            return body.length() <= 1_000 ? body : body.substring(0, 1_000) + "... [truncated]";
        }
    }

    // --- Agent ---
    static class Agent {
        final OllamaClient client;
        final int maxTurns;
        String systemPrompt;
        final List<ToolSpec> toolSpecs = new ArrayList<>();
        final Map<String, Function<OllamaToolCall, String>> toolExecutors = new HashMap<>();
        final List<ChatMessage> pendingMessages = new ArrayList<>();
        final List<String> traceLog = new ArrayList<>();
        long totalPromptEvalCount = 0;
        long totalEvalCount = 0;
        long totalDurationNanos = 0;
        long totalPromptEvalDurationNanos = 0;
        long totalEvalDurationNanos = 0;

        // Hook: called after write_file, returns additional feedback (or null)
        BiFunction<String, String, String> afterWriteHook = null;

        // Hook: called when agent finishes (no more tool calls), can queue more work
        Function<Agent, Boolean> afterDoneHook = null;

        // Permission: safe bash prefixes
        List<String> safeBashPrefixes = List.of(
                "mvn ", "grep ", "find ", "cat ", "ls ", "head ", "tail ", "wc ", "diff ", "git ");

        Agent(int maxTurns) {
            this.client = new OllamaClient();
            this.maxTurns = maxTurns;
            this.systemPrompt = """
                    You are a coding agent operating on the bookshelf/ codebase — a Spring Boot library service.
                    You have tools to read files, write files, run bash commands, grep, and glob.

                    Work methodically:
                    1. Understand the task by reading relevant files first
                    2. Plan your approach before making changes
                    3. Make changes carefully, one file at a time
                    4. Verify your changes compile and tests pass using: cd bookshelf && mvn test -q

                    Be precise. Don't guess at file contents — read them first.
                    When you're done, summarize what you changed and why.
                    """;

            // Register default tools
            registerTool(ReadFile.SPEC, tu -> tu.input(ReadFile.class).execute());
            registerTool(WriteFile.SPEC, tu -> {
                var tool = tu.input(WriteFile.class);
                if (tool.path == null || tool.path.isBlank()) return "Error: path is required.";
                if (tool.content == null) return "Error: content is required.";
                if (!checkPermission("write_file", tool.path)) return "Permission denied by user.";
                var result = tool.execute();
                if (afterWriteHook != null) {
                    var feedback = afterWriteHook.apply(tool.path, tool.content);
                    if (feedback != null) return result + "\n\n--- SENSOR FEEDBACK ---\n" + feedback;
                }
                return result;
            });
            registerTool(RunBash.SPEC, tu -> {
                var tool = tu.input(RunBash.class);
                if (tool.command == null || tool.command.isBlank()) return "Error: command is required.";
                if (!checkPermission("run_bash", tool.command)) return "Permission denied by user.";
                return tool.execute();
            });
            registerTool(GrepFiles.SPEC, tu -> tu.input(GrepFiles.class).execute());
            registerTool(GlobFiles.SPEC, tu -> tu.input(GlobFiles.class).execute());
        }

        void registerTool(ToolSpec toolSpec, Function<OllamaToolCall, String> executor) {
            toolSpecs.add(toolSpec);
            toolExecutors.put(toolSpec.name(), executor);
        }

        void prependSystemPrompt(String extra) {
            systemPrompt = extra + "\n\n" + systemPrompt;
        }

        void addUserMessage(String content) {
            pendingMessages.add(ChatMessage.user(content));
        }

        void run(String task) {
            IO.println("Task: " + task);
            IO.println("Model: " + modelName());
            IO.println();

            var messages = new ArrayList<ChatMessage>();
            messages.add(ChatMessage.system(systemPrompt));
            messages.add(ChatMessage.user(task));

            var turn = 0;
            while (turn < maxTurns) {
                turn++;
                trace("Turn " + turn);
                if (turn == maxTurns - 2) IO.println("[warning] Approaching turn limit (" + maxTurns + ")");
                applyPendingMessages(messages);

                var response = client.chat(messages, toolSpecs);
                recordUsage(response);

                var message = response.message;
                var assistantContent = message.content == null ? "" : message.content;
                if (!assistantContent.isBlank()) IO.println(assistantContent);

                var toolCalls = message.tool_calls == null ? List.<OllamaToolCall>of() : message.tool_calls;
                messages.add(ChatMessage.assistant(assistantContent, toolCalls));

                if (toolCalls.isEmpty()) {
                    if (afterDoneHook != null && afterDoneHook.apply(this)) {
                        trace("afterDoneHook queued more work");
                        continue;
                    }
                    break;
                }

                for (var toolCall : toolCalls) {
                    var toolName = toolCall.name();
                    IO.println("[turn " + turn + "] " + toolName);
                    var result = executeToolCall(toolCall);
                    IO.println("[result] " + truncate(result, 200));
                    messages.add(ChatMessage.tool(toolName, result));
                }
            }

            if (turn >= maxTurns) IO.println("[stopped] Turn limit reached (" + maxTurns + ")");
            IO.println("\n=== Done in " + turn + " turns ===");
            printTraceSummary();
        }

        String complete(String systemPrompt, String userMessage) {
            var messages = List.of(ChatMessage.system(systemPrompt), ChatMessage.user(userMessage));
            var response = client.chat(messages, List.of());
            recordUsage(response);
            return response.message.content == null ? "" : response.message.content;
        }

        void applyPendingMessages(List<ChatMessage> messages) {
            if (pendingMessages.isEmpty()) return;
            messages.addAll(pendingMessages);
            trace("Applied " + pendingMessages.size() + " pending user message(s)");
            pendingMessages.clear();
        }

        String executeToolCall(OllamaToolCall toolCall) {
            var executor = toolExecutors.get(toolCall.name());
            if (executor == null) return "Unknown tool: " + toolCall.name();
            try {
                return executor.apply(toolCall);
            } catch (IllegalArgumentException e) {
                return "Invalid tool arguments: " + e.getMessage();
            } catch (Exception e) {
                return "Tool failed: " + e.getMessage();
            }
        }

        boolean checkPermission(String tool, String detail) {
            if ("run_bash".equals(tool)) {
                for (var prefix : safeBashPrefixes) {
                    if (detail.startsWith(prefix)) return true;
                }
            }
            IO.println("[permission] " + tool + ": " + detail);
            IO.println("  Allow? (y/n): ");
            var response = IO.readln().trim().toLowerCase();
            var allowed = response.startsWith("y");
            trace((allowed ? "ALLOWED" : "DENIED") + ": " + tool + " — " + truncate(detail, 80));
            return allowed;
        }

        void recordUsage(ChatResponse response) {
            if (response.prompt_eval_count != null) totalPromptEvalCount += response.prompt_eval_count;
            if (response.eval_count != null) totalEvalCount += response.eval_count;
            if (response.total_duration != null) totalDurationNanos += response.total_duration;
            if (response.prompt_eval_duration != null) totalPromptEvalDurationNanos += response.prompt_eval_duration;
            if (response.eval_duration != null) totalEvalDurationNanos += response.eval_duration;
            trace("Local usage: +" + valueOrZero(response.prompt_eval_count) + " prompt eval, +"
                    + valueOrZero(response.eval_count) + " eval");
        }

        void trace(String event) {
            traceLog.add("[" + java.time.Instant.now() + "] " + event);
        }

        void printTraceSummary() {
            IO.println("\n=== Trace Summary ===");
            for (var entry : traceLog) IO.println("  " + entry);
            IO.println("  Total prompt eval count: " + totalPromptEvalCount);
            IO.println("  Total eval count:        " + totalEvalCount);
            IO.println("  Total duration:          " + formatDuration(totalDurationNanos));
            IO.println("  Prompt eval duration:    " + formatDuration(totalPromptEvalDurationNanos));
            IO.println("  Eval duration:           " + formatDuration(totalEvalDurationNanos));
            IO.println("  Estimated API cost: $0.0000 (local Ollama)");
        }

        static long valueOrZero(Long value) {
            return value == null ? 0 : value;
        }

        static String formatDuration(long nanos) {
            if (nanos <= 0) return "n/a";
            return String.format("%.3fs", nanos / 1_000_000_000.0);
        }

        static String truncate(String s, int max) {
            if (s == null) return "";
            return s.length() <= max ? s : s.substring(0, max) + "... [truncated]";
        }
    }

    // --- Tools ---

    static class ReadFile {
        static final ToolSpec SPEC = new ToolSpec(
                "read_file",
                "Read the contents of a file. Always read a file before modifying it.",
                objectSchema(Map.of("path", stringProperty("Absolute or relative file path to read")), List.of("path")));

        public String path;

        public String execute() {
            if (path == null || path.isBlank()) return "Error: path is required.";
            try { return Files.readString(Path.of(path)); }
            catch (IOException e) { return "Error: " + e.getMessage(); }
        }
    }

    static class WriteFile {
        static final ToolSpec SPEC = new ToolSpec(
                "write_file",
                "Write content to a file. Creates parent directories if needed. Always read the file first.",
                objectSchema(Map.of(
                        "path", stringProperty("File path to write to"),
                        "content", stringProperty("The complete file content to write")),
                        List.of("path", "content")));

        public String path;
        public String content;

        public String execute() {
            if (path == null || path.isBlank()) return "Error: path is required.";
            if (content == null) return "Error: content is required.";
            try {
                var p = Path.of(path);
                if (p.getParent() != null) Files.createDirectories(p.getParent());
                Files.writeString(p, content);
                return "Wrote " + content.length() + " chars to " + path;
            } catch (IOException e) { return "Error: " + e.getMessage(); }
        }
    }

    static class RunBash {
        static final ToolSpec SPEC = new ToolSpec(
                "run_bash",
                "Run a bash command. Output is capped at 10k chars.",
                objectSchema(Map.of("command", stringProperty("The bash command to run")), List.of("command")));

        public String command;

        public String execute() {
            if (command == null || command.isBlank()) return "Error: command is required.";
            try {
                var proc = new ProcessBuilder("bash", "-c", command).redirectErrorStream(true).start();
                var out = new String(proc.getInputStream().readAllBytes());
                var code = proc.waitFor();
                var result = out.length() > MAX_OUTPUT ? out.substring(0, MAX_OUTPUT) + "\n...[truncated]" : out;
                return result + "\n[exit code: " + code + "]";
            } catch (Exception e) { return "Error: " + e.getMessage(); }
        }
    }

    static class GrepFiles {
        static final ToolSpec SPEC = new ToolSpec(
                "grep_files",
                "Search file contents with grep. Returns matching lines with file:line prefixes.",
                objectSchema(Map.of(
                        "pattern", stringProperty("Regex pattern to search for"),
                        "path", stringProperty("Directory or file to search in")),
                        List.of("pattern", "path")));

        public String pattern;
        public String path;

        public String execute() {
            if (pattern == null || pattern.isBlank()) return "Error: pattern is required.";
            if (path == null || path.isBlank()) return "Error: path is required.";
            try {
                var proc = new ProcessBuilder("grep", "-rn", "--include=*.java", pattern, path)
                        .redirectErrorStream(true).start();
                var out = new String(proc.getInputStream().readAllBytes());
                proc.waitFor();
                return out.length() > MAX_OUTPUT ? out.substring(0, MAX_OUTPUT) + "\n...[truncated]"
                        : out.isEmpty() ? "No matches." : out;
            } catch (Exception e) { return "Error: " + e.getMessage(); }
        }
    }

    static class GlobFiles {
        static final ToolSpec SPEC = new ToolSpec(
                "glob_files",
                "Find files by glob pattern.",
                objectSchema(Map.of(
                        "pattern", stringProperty("Glob pattern, e.g. **/*.java"),
                        "path", stringProperty("Base directory to search from")),
                        List.of("pattern", "path")));

        public String pattern;
        public String path;

        public String execute() {
            if (pattern == null || pattern.isBlank()) return "Error: pattern is required.";
            if (path == null || path.isBlank()) return "Error: path is required.";
            try {
                var base = Path.of(path);
                var matcher = FileSystems.getDefault().getPathMatcher("glob:" + pattern);
                try (var stream = Files.walk(base)) {
                    var matches = stream.filter(Files::isRegularFile)
                            .filter(p -> matcher.matches(base.relativize(p)))
                            .map(Path::toString).collect(Collectors.joining("\n"));
                    return matches.isEmpty() ? "No files matched." : matches;
                }
            } catch (IOException e) { return "Error: " + e.getMessage(); }
        }
    }
}

File: python/steps/inner_harness.py
"""Inner harness -- shared library for Ollama-based local coding agents.

Provides the Agent class with built-in tools, permission system, hooks,
and usage tracking. Steps 4-7 import from this module.
"""

import json
import os
import subprocess
import sys
from datetime import datetime, timezone
from pathlib import Path

import httpx

DEFAULT_MODEL = "qwen3:8b"
DEFAULT_BASE_URL = "http://localhost:11434"
MAX_OUTPUT = 10_000


# --- Config ---

def model_name():
    return (os.environ.get("OLLAMA_MODEL") or DEFAULT_MODEL).strip()


def base_url():
    return (os.environ.get("OLLAMA_BASE_URL") or DEFAULT_BASE_URL).strip().rstrip("/")


def setup_hint():
    return (
        f"\nOllama setup:\n"
        f"  ollama serve\n"
        f"  ollama pull {model_name()}\n"
    )


# --- Schema helpers ---

def string_property(description):
    return {"type": "string", "description": description}


def object_schema(properties, required):
    return {
        "type": "object",
        "properties": properties,
        "required": required,
    }


# --- Ollama chat ---

def chat(messages, tools):
    body = {
        "model": model_name(),
        "stream": False,
        "messages": messages,
        "tools": tools,
    }
    try:
        with httpx.Client(timeout=300) as client:
            resp = client.post(f"{base_url()}/api/chat", json=body)
    except httpx.ConnectError:
        raise SystemExit(f"Could not connect to Ollama at {base_url()}.{setup_hint()}")
    except httpx.HTTPError as e:
        raise SystemExit(f"Could not call Ollama at {base_url()}: {e}{setup_hint()}")

    if resp.status_code < 200 or resp.status_code >= 300:
        raise SystemExit(
            f"Ollama chat failed with HTTP {resp.status_code}: {resp.text}{setup_hint()}"
        )

    data = resp.json()
    if data.get("error"):
        raise SystemExit(f"Ollama chat failed: {data['error']}{setup_hint()}")
    return data


# --- Utility ---

def truncate(text, max_len=200):
    if not text:
        return ""
    if len(text) <= max_len:
        return text
    return text[:max_len] + "... [truncated]"


def format_nanos(nanos):
    if nanos <= 0:
        return "n/a"
    return f"{nanos / 1_000_000_000:.3f}s"


# --- Agent ---

class Agent:
    """Ollama-based coding agent with tools, hooks, and usage tracking."""

    SAFE_BASH_PREFIXES = [
        "pytest ", "grep ", "find ", "cat ", "ls ",
        "head ", "tail ", "wc ", "diff ", "git ",
    ]

    DEFAULT_SYSTEM_PROMPT = """\
You are a coding agent operating on the bookshelf/ codebase -- a FastAPI library service.
You have tools to read files, write files, run bash commands, grep, and glob.

Work methodically:
1. Understand the task by reading relevant files first
2. Plan your approach before making changes
3. Make changes carefully, one file at a time
4. Verify your changes by running: cd bookshelf && python -m pytest -q

Be precise. Don't guess at file contents -- read them first.
When you're done, summarize what you changed and why."""

    def __init__(self, max_turns=25):
        self.max_turns = max_turns
        self.system_prompt = self.DEFAULT_SYSTEM_PROMPT
        self.messages: list[dict] = []
        self.tools: dict[str, dict] = {}  # name -> {"schema": ..., "handler": ...}
        self.pending_messages: list[dict] = []
        self.trace_log: list[str] = []
        self.total_prompt_eval_count = 0
        self.total_eval_count = 0
        self.total_duration_ns = 0
        self.total_prompt_eval_duration_ns = 0
        self.total_eval_duration_ns = 0

        # Hooks (default None)
        self.after_write_hook = None   # (file_path, content) -> str | None
        self.after_done_hook = None    # (agent) -> bool

        # Register default tools
        self._register_default_tools()

    def _register_default_tools(self):
        self.register_tool(
            "read_file",
            "Read the contents of a file. Always read a file before modifying it.",
            {"path": string_property("Absolute or relative file path to read")},
            self._tool_read_file,
        )
        self.register_tool(
            "write_file",
            "Write content to a file. Creates parent directories if needed. "
            "Always read the file first to understand existing content.",
            {"path": string_property("File path to write to"),
             "content": string_property("The complete file content to write")},
            self._tool_write_file,
        )
        self.register_tool(
            "run_bash",
            "Run a bash command. Use for: pytest, pip, git diff, etc. "
            "Output is capped at 10k chars.",
            {"command": string_property("The bash command to run")},
            self._tool_run_bash,
        )
        self.register_tool(
            "grep_files",
            "Search file contents with grep. Returns matching lines with file:line prefixes.",
            {"pattern": string_property("Regex pattern to search for"),
             "path": string_property("Directory or file to search in")},
            self._tool_grep_files,
        )
        self.register_tool(
            "glob_files",
            "Find files by glob pattern.",
            {"pattern": string_property("Glob pattern, e.g. **/*.py"),
             "path": string_property("Base directory to search from")},
            self._tool_glob_files,
        )

    def register_tool(self, name, description, parameters, handler):
        """Register a tool with its schema and handler function.

        parameters: dict of property_name -> property_schema
        handler: callable(arguments_dict) -> str
        """
        schema = {
            "type": "function",
            "function": {
                "name": name,
                "description": description,
                "parameters": object_schema(
                    parameters,
                    list(parameters.keys()),
                ),
            },
        }
        self.tools[name] = {"schema": schema, "handler": handler}

    def prepend_system_prompt(self, text):
        self.system_prompt = text + "\n\n" + self.system_prompt

    def add_user_message(self, text):
        self.pending_messages.append({"role": "user", "content": text})

    def complete(self, system_prompt, user_message):
        """One-shot completion (no tools)."""
        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_message},
        ]
        response = chat(messages, [])
        self._record_usage(response)
        msg = response.get("message")
        return (msg.get("content") or "") if msg else ""

    def run(self, task):
        """Main agent loop: add system + user messages, loop with tool calls."""
        print(f"Task: {task}")
        print(f"Model: {model_name()}")
        print()

        messages = [
            {"role": "system", "content": self.system_prompt},
            {"role": "user", "content": task},
        ]

        turn = 0
        while turn < self.max_turns:
            turn += 1
            self._trace(f"Turn {turn}")
            if turn == self.max_turns - 2:
                print(f"[warning] Approaching turn limit ({self.max_turns})")

            # Apply any pending messages
            self._apply_pending(messages)

            tool_schemas = [t["schema"] for t in self.tools.values()]
            response = chat(messages, tool_schemas)
            self._record_usage(response)

            assistant = response.get("message")
            if assistant is None:
                raise SystemExit(
                    f"Ollama response did not include a message.{setup_hint()}"
                )

            content = (assistant.get("content") or "").strip()
            if content:
                print(content)

            tool_calls = assistant.get("tool_calls") or []
            messages.append({
                "role": "assistant",
                "content": assistant.get("content"),
                "tool_calls": tool_calls if tool_calls else None,
            })

            if not tool_calls:
                if self.after_done_hook is not None and self.after_done_hook(self):
                    self._trace("after_done_hook queued more work")
                    continue
                break

            for tc in tool_calls:
                fn = tc.get("function", {})
                name = fn.get("name", "")
                arguments = fn.get("arguments", {})
                if isinstance(arguments, str):
                    arguments = json.loads(arguments)
                print(f"[turn {turn}] {name}")
                result = self._execute_tool(name, arguments)
                print(f"[result] {truncate(result)}")
                messages.append({"role": "tool", "content": result})

        if turn >= self.max_turns:
            print(f"[stopped] Turn limit reached ({self.max_turns})")
        print(f"\n=== Done in {turn} turns ===")
        self._print_trace_summary()

    # --- Internal ---

    def _apply_pending(self, messages):
        if not self.pending_messages:
            return
        messages.extend(self.pending_messages)
        self._trace(f"Applied {len(self.pending_messages)} pending user message(s)")
        self.pending_messages.clear()

    def _execute_tool(self, name, arguments):
        tool = self.tools.get(name)
        if tool is None:
            return f"Unknown tool: {name}"
        try:
            return tool["handler"](arguments)
        except Exception as e:
            return f"Tool failed: {e}"

    def _check_permission(self, tool, detail):
        if tool == "run_bash":
            for prefix in self.SAFE_BASH_PREFIXES:
                if detail.startswith(prefix):
                    return True
        print(f"[permission] {tool}: {detail}")
        answer = input("  Allow? (y/n): ").strip().lower()
        allowed = answer.startswith("y")
        self._trace(f"{'ALLOWED' if allowed else 'DENIED'}: {tool} -- {truncate(detail, 80)}")
        return allowed

    def _record_usage(self, response):
        self.total_prompt_eval_count += response.get("prompt_eval_count", 0)
        self.total_eval_count += response.get("eval_count", 0)
        self.total_duration_ns += response.get("total_duration", 0)
        self.total_prompt_eval_duration_ns += response.get("prompt_eval_duration", 0)
        self.total_eval_duration_ns += response.get("eval_duration", 0)
        self._trace(
            f"Usage: +{response.get('prompt_eval_count', 0)} prompt eval, "
            f"+{response.get('eval_count', 0)} eval tokens"
        )

    def _trace(self, event):
        timestamp = datetime.now(timezone.utc).isoformat()
        self.trace_log.append(f"[{timestamp}] {event}")

    def _print_trace_summary(self):
        print("\n=== Trace Summary ===")
        for entry in self.trace_log:
            print(f"  {entry}")
        print(f"  Total prompt eval count:       {self.total_prompt_eval_count}")
        print(f"  Total eval count:              {self.total_eval_count}")
        print(f"  Total duration:                {format_nanos(self.total_duration_ns)}")
        print(f"  Total prompt eval duration:    {format_nanos(self.total_prompt_eval_duration_ns)}")
        print(f"  Total eval duration:           {format_nanos(self.total_eval_duration_ns)}")
        print("  Estimated API cost: $0.0000 (local Ollama)")

    # --- Default tool implementations ---

    def _tool_read_file(self, arguments):
        path = arguments.get("path", "")
        if not path:
            return "Error: path is required."
        try:
            with open(path) as f:
                return f.read()
        except OSError as e:
            return f"Error: {e}"

    def _tool_write_file(self, arguments):
        path = arguments.get("path", "")
        content = arguments.get("content")
        if not path:
            return "Error: path is required."
        if content is None:
            return "Error: content is required."
        if not self._check_permission("write_file", path):
            self._trace(f"DENIED: write_file -- {path}")
            return "Permission denied by user."
        self._trace(f"ALLOWED: write_file -- {path}")
        try:
            p = Path(path)
            p.parent.mkdir(parents=True, exist_ok=True)
            p.write_text(content)
            result = f"Wrote {len(content)} chars to {path}"
        except OSError as e:
            return f"Error: {e}"

        # After-write hook
        if self.after_write_hook is not None:
            feedback = self.after_write_hook(path, content)
            if feedback is not None:
                return result + "\n\n--- SENSOR FEEDBACK ---\n" + feedback
        return result

    def _tool_run_bash(self, arguments):
        command = arguments.get("command", "")
        if not command:
            return "Error: command is required."
        if not self._check_permission("run_bash", command):
            self._trace(f"DENIED: run_bash -- {command}")
            return "Permission denied by user."
        self._trace(f"ALLOWED: run_bash -- {command}")
        try:
            proc = subprocess.run(
                ["bash", "-c", command],
                capture_output=True, text=True, timeout=120,
            )
            output = proc.stdout + proc.stderr
            if len(output) > MAX_OUTPUT:
                output = output[:MAX_OUTPUT] + "\n...[truncated]"
            return output + f"\n[exit code: {proc.returncode}]"
        except Exception as e:
            return f"Error: {e}"

    def _tool_grep_files(self, arguments):
        pattern = arguments.get("pattern", "")
        path = arguments.get("path", "")
        if not pattern:
            return "Error: pattern is required."
        if not path:
            return "Error: path is required."
        try:
            proc = subprocess.run(
                ["grep", "-rn", "--include=*.py", pattern, path],
                capture_output=True, text=True, timeout=30,
            )
            output = proc.stdout + proc.stderr
            if len(output) > MAX_OUTPUT:
                return output[:MAX_OUTPUT] + "\n...[truncated]"
            return output if output else "No matches."
        except Exception as e:
            return f"Error: {e}"

    def _tool_glob_files(self, arguments):
        pattern = arguments.get("pattern", "")
        path = arguments.get("path", "")
        if not pattern:
            return "Error: pattern is required."
        if not path:
            return "Error: path is required."
        try:
            matches = sorted(str(p) for p in Path(path).glob(pattern) if p.is_file())
            return "\n".join(matches) if matches else "No files matched."
        except Exception as e:
            return f"Error: {e}"


# --- Self-test when run directly ---

if __name__ == "__main__":
    task = (
        " ".join(sys.argv[1:])
        if len(sys.argv) > 1
        else "Read bookshelf/pyproject.toml and tell me what project this is."
    )
    agent = Agent(max_turns=25)
    agent.run(task)
